smaug/python/unique_name_test.py

```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env python
+
+""" This tests the unique name creation in the Graph class."""
+
+import unittest
+import numpy as np
+
+from smaug.core import types_pb2
+from smaug.python.tensor import Tensor
+from smaug.python.graph import Graph
+from smaug.python.ops import math_ops
+
+graph_name = "test_graph"
+backend = "Reference"
+x = Tensor(
+    data_layout=types_pb2.N, tensor_data=np.random.rand(4).astype(np.float32))
+y = Tensor(
+    data_layout=types_pb2.N, tensor_data=np.random.rand(4).astype(np.float32))
+
+def get_node_names(graph):
+  return set([node.name for node in graph.get_nodes()])
+
+class TestUniqueName(unittest.TestCase):
+  def test_default_names(self):
+    with Graph(graph_name, backend) as test_graph:
+      res= math_ops.add(x, y)
+      res = math_ops.mul(x, res)
+    self.assertEqual(get_node_names(test_graph), {"add", "mul"})
+
+  def test_auto_unique_names(self):
+    with Graph(graph_name, backend) as test_graph:
+      res = math_ops.add(x, y)
+      res = math_ops.add(res, res)
+      res = math_ops.add(res, res)
+    self.assertEqual(get_node_names(test_graph), {"add", "add_1", "add_2"})
+
+  def test_user_supplied_names0(self):
+    with Graph(graph_name, backend) as test_graph:
+      res = math_ops.add(x, y, name="add")
+      res = math_ops.mul(res, res, name="mul")
+    self.assertEqual(get_node_names(test_graph), {"add", "mul"})
+
+  def test_user_supplied_names1(self):
+    with Graph(graph_name, backend) as test_graph:
+      res = math_ops.add(x, y, name="add")
+      res = math_ops.add(res, res, name="add_1")
+      res = math_ops.add(res, res, name="add")
+    self.assertEqual(get_node_names(test_graph), {"add", "add_1", "add_2"})
+
+  def test_user_supplied_names1(self):
+    with Graph(graph_name, backend) as test_graph:
+      res = math_ops.add(x, y, name="add")
+      res = math_ops.add(res, res, name="add_3")
+      res = math_ops.add(res, res, name="add")
+    self.assertEqual(get_node_names(test_graph), {"add", "add_1", "add_3"})
+
+  def test_user_supplied_names2(self):
+    with Graph(graph_name, backend) as test_graph:
+      res = math_ops.add(x, y, name="add")
+      res = math_ops.add(res, res, name="add_1")
+      res = math_ops.add(res, res, name="add_1")
+    self.assertEqual(get_node_names(test_graph), {"add", "add_1", "add_1_1"})
+
+  def test_user_supplied_names3(self):
+    with Graph(graph_name, backend) as test_graph:
+      res = math_ops.add(x, y, name="add")
+      unique_name = test_graph.create_unique_name("add", mark_as_used=False)
```

Comment:
Can we accomplish testing without having to add a `mark_as_used` parameter and making the naming here public? It seems very unusual to add this parameter. Why can't we just compare the name of the operator with what we expect? What is the benefit of exposing `create_unique_name`?

Reply:
Yeah, I admit by now we don't have a use case for this. But it's needed (believe me) in an upcoming commit for implementing the conditional operator, where I need to know the name of the operator right before I create it, so I can use the name to avoid some deadlock.

Reply:
Okay, I'm very curious to see this use case.